### PR TITLE
6/5 Feature/nijo 787 change booking to work with third step

### PR DESCRIPTION
**Pull request by mingazhev**

### CODE REVIEW

File: Api/Models/Bookings/AccommodationBookingRequest.cs [hunk 1]
```diff
@@ -55,12 +57,17 @@ namespace HappyTravel.Edo.Api.Models.Bookings
...
+        public string AvailabilityId { get; }
+
         /// <summary>
         ///     The residency of a main passenger.
         /// </summary>
         [Required]
         public string Residency { get; }
 
+        public Guid SearchId { get; }
+        public Guid ResultId { get; }
```
Comment (kirillta) [inline Api/Models/Bookings/AccommodationBookingRequest.cs:69]: What `ResultId` means?

File: Api/Services/Accommodations/Bookings/BookingService.cs [hunk 2]
```diff
@@ -29,45 +31,51 @@
...
         {
-            var (_, isCachedAvailabilityFailure, responseWithMarkup, cachedAvailabilityError) = await _availabilityResultsCache.Get(bookingRequest.DataProvider, bookingRequest.AvailabilityId);
+            var (_, isCachedAvailabilityFailure, responseWithMarkup, cachedAvailabilityError) = await _bookingEvaluationResultsStorage.Get(bookingRequest.SearchId, bookingRequest.ResultId, bookingRequest.RoomContractSetId, _dataProviderOptions.EnabledProviders);
             if (isCachedAvailabilityFailure)
                 return ProblemDetailsBuilder.Fail<string>(cachedAvailabilityError);
             
-            var bookingAvailability = ExtractBookingAvailabilityInfo(responseWithMarkup.Data);
+            var bookingAvailability = ExtractBookingAvailabilityInfo(responseWithMarkup.Source, responseWithMarkup.Result.Data);
+            // Temporarily saving availability id along with booking request to get it on the booking step.
+            // TODO: Rewrite this to save such data in another place
```
Comment (kirillta) [inline Api/Services/Accommodations/Bookings/BookingService.cs:77]: Would you add a task for that?

File: Api/Services/Accommodations/Availability/Steps/BookingEvaluation/BookingEvaluationResultsStorage.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+        public Task Set(Guid searchId, Guid resultId, Guid roomContractSetId, DataWithMarkup<SingleAccommodationAvailabilityDetailsWithDeadline> availability,
+            DataProviders dataProvider)
+        {
+            var key = CreateKeyPrefix(searchId, resultId, roomContractSetId);
+            return _availabilityStorage.SaveObject(key, availability, dataProvider);
+        }
+
+
+        public Task<Result<(DataProviders Source, DataWithMarkup<SingleAccommodationAvailabilityDetailsWithDeadline> Result)>> Get(Guid searchId, Guid resultId, Guid roomContractSetId, List<DataProviders> dataProviders) => throw new NotImplementedException();
```
Comment (kirillta) [inline Api/Services/Accommodations/Availability/Steps/BookingEvaluation/BookingEvaluationResultsStorage.cs:27]: Is that throwing intentional?

File: Api/Models/Bookings/AccommodationBookingRequest.cs (hunk 1 above)
Comment (mingazhev) [reply Api/Models/Bookings/AccommodationBookingRequest.cs:69]: I used following names:
- search is for the whole search (wide)
- result is the selected result from first step (accommodation)
- room contract set is the selected room contract set from the second step

File: Api/Services/Accommodations/Bookings/BookingService.cs (hunk 2 above)
Comment (mingazhev) [reply Api/Services/Accommodations/Bookings/BookingService.cs:77]: I've added now, NIJO-813

File: Api/Services/Accommodations/Availability/Steps/BookingEvaluation/BookingEvaluationResultsStorage.cs (hunk 3 above)
Comment (mingazhev) [reply Api/Services/Accommodations/Availability/Steps/BookingEvaluation/BookingEvaluationResultsStorage.cs:27]: No it wasn't. I've finished implementation now

File: Api/Services/Accommodations/Availability/Steps/RoomSelection/RoomSelectionStorage.cs [hunk 4]
```diff
@@ -21,15 +21,17 @@ public Task SaveResult(Guid searchId, Guid resultId, SingleAccommodationAvailabi
             return _storage.SaveObject(keyPrefix, details, dataProvider);
         }
         
-        public Task<(DataProviders DataProvider, SingleAccommodationAvailabilityDetails Result)[]> GetResult(Guid searchId, Guid resultId, List<DataProviders> dataProviders)
+        public async Task<(DataProviders DataProvider, SingleAccommodationAvailabilityDetails Result)[]> GetResult(Guid searchId, Guid resultId, List<DataProviders> dataProviders)
```
Comment (kirillta) [inline Api/Services/Accommodations/Availability/Steps/RoomSelection/RoomSelectionStorage.cs:24]: Why an array?

**Pull request merged**